services/src/ogc/wms/request.rs

```diff
@@ -0,0 +1,199 @@
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetMap {
+    pub version: String,
+    #[serde(deserialize_with = "from_str")]
+    pub width: u32,
+    #[serde(deserialize_with = "from_str")]
+    pub height: u32,
+    #[serde(deserialize_with = "parse_bbox")]
+    pub bbox: BoundingBox2D,
+    pub format: GetMapFormat,
+    pub layer: String,
+    pub crs: String,  // TODO: parse CRS
+    pub styles: String,
+    #[serde(default)]
+    #[serde(deserialize_with = "parse_time")]
+    pub time: Option<TimeInterval>,
+    // TODO: parse Option<TimeInterval>,
+    pub transparent: Option<bool>,
+    pub bgcolor: Option<String>,
+    pub sld: Option<String>,
+    pub sld_body: Option<String>,
+    pub elevation: Option<String>,
+    pub exceptions: Option<String>, // TODO: parse Option<GetMapExceptionFormat>
+    // DIM_<name>
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetMapExceptionFormat {
+    TextXML //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetMapFormat {
+    #[serde(rename = "image/png")]
+    ImagePng //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetFeatureInfo {
+    pub version: String,
+    pub query_layers: String,
+    pub info_format: Option<String>, // TODO: parse Option<GetFeatureInfoFormat>,
+    // ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetFeatureInfoFormat {
+    TextXML //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetStyles {
+    pub version: String,
+    pub layer: String,
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetLegendGraphic {
+    pub version: String,
+    pub layer: String,
+    // ...
+}
+
+pub fn parse_bbox<'de, D>(deserializer: D) -> Result<BoundingBox2D, D::Error>
+    where D: serde::Deserializer<'de>
+{
+    // format is : "x1,y1,x2,y2"
+
+    let s = <&str as serde::Deserialize>::deserialize(deserializer)?;
+
+    let split: Vec<Result<f64, std::num::ParseFloatError>> = s.split(',')
+        .map(str::parse).collect();
+
+    if let [Ok(x1), Ok(y1), Ok(x2), Ok(y2)] = *split.as_slice() {
+        BoundingBox2D::new(
+            Coordinate2D::new(x1, y1),
+            Coordinate2D::new(x2, y2))
+            .map_err(|_| D::Error::custom("Not a valid bbox"))
+    } else {
+        Err(D::Error::custom("Not a valid bbox"))
+    }
+}
+
+pub fn parse_time<'de, D>(deserializer: D) -> Result<Option<TimeInterval>, D::Error>
+    where D: serde::Deserializer<'de>
+{
+    // time is specified in ISO8601, it can either be an instant (single datetime) or an interval
+    // An interval is separated by "/". "Either the start value or the end value can be omitted to
+    // indicate no restriction on time in that direction."
+    // sources: - http://docs.geoserver.org/2.8.x/en/user/services/wms/time.html#wms-time
+    //          - http://www.ogcnetwork.net/node/178
+
+    // TODO: support relative time intervals
```

Comment:
Should be in the docstring

Reply:
d9fdc9060ef49ea9397e10d0958e5efc00716bd8